src/ray/common/id.cc

```diff
@@ -194,6 +194,18 @@ TaskID TaskID::ForNormalTask(const JobID &job_id, const TaskID &parent_task_id,
   return TaskID::FromBinary(data);
 }
 
+TaskID TaskID::ForExecution(const TaskID &task_id, uint64_t execution_counter) {
+  std::string data_str;
+  std::copy_n(task_id.Data(), TaskID::kLength, std::back_inserter(data_str));
+  static_assert(TaskID::kUniqueBytesLength >= 8, "TaskID must have at least 64 bits");
+  auto data = reinterpret_cast<uint64_t *>(data_str.data());
+  // Zero out the low byte for readability.
+  (*reinterpret_cast<uint8_t *>(data)) = 0;
+  // Add execution_counter to the task ID unique bytes.
+  (*data) += execution_counter;
```

Comment:
Do we need to consider endianness here?

Reply:
The correctness should be fine because casting to `uint8_t` will truncate the higher-order bits. But the endianness affects which bytes get zero-ed out (on my laptop the first two bytes of the .hex() representation are zeroed).

Reply:
I think I'm worried about the ` (*data) += execution_counter;` like are we always adding the counter to the bytes that are zero-ed out.

Reply:
I'm not sure if I understand - the intention is to add the counter to the zeroed out bytes.

Reply:
```
#include <stdio.h>
#include <stdint.h>
int main() {
  uint16_t bytes = 1;
  if (*((uint8_t*) &bytes) == 1) {
    printf("Detected Host Little Endian.\n");
  } else {
    printf("Detected Host Big Endian.\n");
  }
  uint8_t data[4];
  int i = 0;
  for (i = 0; i < 4; ++i) {
    data[i] = 1;
  }
  uint32_t *ptr = (uint32_t *)data;
  (*(uint8_t *)(ptr)) = 0;
  (*ptr) += 1;
  for (i = 0; i < 4; ++i) {
    printf("%d\n", data[i]);
  }
}
```

```
root@f2c9044927aa:/# ./a.out 
Detected Host Big Endian.
0
1
1
2
```

```
(ray) jjyao@Jiajuns-MacBook-Pro:~/Workspace/ray1 jjyao/parallelism$ ./a.out 
Detected Host Little Endian.
1
1
1
1
```

Reply:
Ah I see ,my bad! I'm just going to change this to use an extra uint8_t field for the counter for simplicity/portability.

Reply:
Actually just think of a portable way, instead of `(*reinterpret_cast<uint8_t *>(data)) = 0;` we can do `*data & 0xFFFFFF...0`

Reply:
Ah good idea, thanks!